src/Ouroboros/Network/Protocol/Get/Client.hs

```diff
@@ -0,0 +1,28 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE RecordWildCards #-}
+
+module Ouroboros.Network.Protocol.Get.Client where
+
+import Ouroboros.Network.Protocol.Typed
+import Ouroboros.Network.Protocol.Get.Type
+
+data Client m resource resourceId a = Client {
+    -- | Handle incoming data.
+    handleData :: Maybe resource -> m a,
+    -- | Request resource indenfied by @'getResourceId'@
+    getResourceId :: resourceId
+  }
```

Comment:
This type has a fixed resource identifier, so always requests the same thing. The style I endorse is something like

```Haskell
data Client m resource resourceId a where
  Request :: resourceId -> (Maybe resource -> m (Client m resource resourceId a)
  Done :: a -> Client m resource resourceId a
```

Reply:
Nice :)

Reply:
The complementary server type can be mechanically derived

```Haskell
-- The client is an additive disjunction (typical sum type), so the server is an additive conjunction.
-- Haskell doesn't have this, so we use a multiplicative conjunction and rely on discipline: use
-- one and only one of the fields.
data Server m resource resourceId a = Server
  { request :: resourceId -> m (Maybe resource, Server m resource resourceId a)
  , done :: a
  }
```

Reply:
@avieth it seems like you're asking for some linear types :)

Reply:
I would love some linear types!

Reply:
btw, we don't need `Done :: Client ...` and then your transformation is just a renaming, but I'm fine with keeping GADTs syntax.

Reply:
If we get rid of `Done` then we should also dump the `a` type parameter, but this means that a client can never stop normally.

It's also not a renaming: mine gives a new `Client` in the `Maybe resource ->` continuation.

Reply:
This protocol stops after receiving a response from the server, when we want more than a single block we should use a streaming protocol, isn't it?

Reply:
> This protocol stops after receiving a response from the server,

Oh you're right, my mistake.

Reply:
But it might makes sense to change the protocol to have a continuation, for example it allows to download a block from some peer when a some mutation variable changes or under some other conditions.  It might be used not just for blocks, so its probably better to keep it extensible.  But there is no close bearer message: is this a problem? adding it will degrade the case of downloading a single resource (that's why I came up with this simple protocol).